### PR TITLE
Allow receipt signers to invalidate receipts

### PR DESCRIPTION
**Pull request by maxsam4**
### new features

- Receipt issuers for settlement can now invalidate and revalidate their receipts

### CODE REVIEW

File: pallets/runtime/tests/src/settlement_test.rs [hunk 1]
```diff
@@ -917,6 +917,35 @@ fn claiming_receipt() {
...
+                    ReceiptDetails {
+                        receipt_uid: 0,
+                        leg_id: 0,
+                        signer: AccountKeyring::Alice.public(),
+                        signature: OffChainSignature::from(
+                            AccountKeyring::Alice.sign(&msg.encode())
+                        ),
+                        metadata: metadata.clone()
+                    }
+                ),
```
Comment (Centril) [inline pallets/runtime/tests/src/settlement_test.rs:940]: Delta to above call is only the metadata, let's dedup with closure.

File: pallets/runtime/tests/src/settlement_test.rs [hunk 2]
```diff
@@ -917,6 +917,35 @@ fn claiming_receipt() {
...
+                Error::ReceiptAlreadyClaimed
+            );
+            assert_ok!(Settlement::change_receipt_validity(
+                alice_signed.clone(),
+                0,
+                true
+            ));
+
             // Claiming, unclaiming and claiming receipt
             assert_ok!(Settlement::claim_receipt(
```
Comment (Centril) [inline pallets/runtime/tests/src/settlement_test.rs:950]: Same here

**Review by Centril — CHANGES_REQUESTED** [review]

File: pallets/settlement/src/lib.rs [hunk 3]
```diff
@@ -775,6 +778,22 @@ decl_module! {
...
+        /// * `receipt_uid` - Unique ID of the receipt.
+        /// * `validity` - New validity of the receipt.
+        #[weight = <T as Trait>::WeightInfo::change_receipt_validity()]
+        pub fn change_receipt_validity(origin, receipt_uid: u64, validity: bool) {
+            let PermissionedCallOriginData {
+                primary_did,
+                sender: signer,
+                ..
+            } = Identity::<T>::ensure_origin_call_permissions(origin)?;
+            <ReceiptsUsed<T>>::insert(&signer, receipt_uid, !validity);
```
Comment (vkomenda) [inline pallets/settlement/src/lib.rs:793]: Why the validity is stored in a negated form while it is emitted in the original form in the event?

I think it should be stored in the original form as well.

Comment (maxsam4) [reply pallets/settlement/src/lib.rs:793]: The storage is for `ReceiptsUsed`, not for Validity. By default, all receipts are unused and we treat unused receipts as Valid.

I didn't want to do a storage migration to change the meaning.

File: pallets/settlement/src/benchmarking.rs [hunk 4]
```diff
@@ -863,6 +863,13 @@ benchmarks! {
         }
     }
 
+    change_receipt_validity {
+        let signer = user::<T>("signer", 0);
+    }: _(signer.origin(), 0, false)
+    verify {
+        assert!(Module::<T>::receipts_used(&signer.account(), 0));
```
Comment (vkomenda) [inline pallets/settlement/src/benchmarking.rs:870]: You can use ensure here with an error message just in case something fails in compiled WASM.

**Review by satyamakgec — APPROVED** [review]

**Review by Centril — APPROVED** [review]

**Review by vkomenda — APPROVED** [review]

**Pull request merged**